File: solutions/lua/clock/1/clock.lua
local clock = {}

function clock.at(hour, minute)
    local n = {
        minute = 0,
        plus = clock.plus,
        minus = clock.minus,
        equals = clock.equals,
        __tostring = clock.__tostring
    }

    hour = (hour or 0)
    minute = (minute or 0)
    n.minute = (hour * 60 + minute) % (24 * 60)

    setmetatable(n, clock)
    return n
end

function clock:plus(n)
    self.minute = self.minute + n
    return self
end

function clock:minus(n)
    self.minute = self.minute - n
    return self
end

function clock:equals(clk)
    local result
    result = (self.minute == clk.minute)
    return result
end

function clock:__tostring()
    return string.format("%02d:%02d", math.floor(self.minute / 60) % 24,
                         self.minute % 60)
end

return clock

File: solutions/cpp/clock/1/clock.cpp
#include "clock.h"
#include <array>
#include <cstdio>
#include <string>

namespace date_independent {

clock::clock(int minute) { minutes = minute; }

clock clock::at(const int hour, const int minute) {
  int min = ((hour * 60) + minute) % (24 * 60);
  min = min < 0 ? (24 * 60) + min : min;
  return clock(min);
}

clock clock::plus(const int minute) {
  int min = (this->minutes + minute) % (24 * 60);
  min = min < 0 ? (24 * 60) + min : min;
  return clock(min);
}

clock::operator std::string() const {
  std::array<char, 10> str{};
  std::sprintf(str.begin(), "%02d:%02d", minutes / 60, minutes % 60);
  return str.data();
}

bool clock::operator==(const clock clk) const {
  return this->minutes == clk.minutes;
}
bool clock::operator!=(const clock clk) const {
  return this->minutes != clk.minutes;
}
} // namespace date_independent
